### PR TITLE
Add `yii\filters\HostControl::$fallbackHostInfo`

### PR DESCRIPTION
**Pull request by rob006**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | https://github.com/yiisoft/yii2/issues/13050#issuecomment-264197674

### CODE REVIEW

**Review by samdark — APPROVED** [review]

File: framework/web/Request.php [hunk 1]
```diff
@@ -547,9 +547,6 @@ public function getQueryParam($name, $defaultValue = null)
      * > If you don't have access to the server configuration, you can setup [[\yii\filters\HostControl]] filter at
      * > application level in order to protect against such kind of attack.
      *
-     * @property string|null schema and hostname part (with port number if needed) of the request URL
-     * (e.g. `http://www.yiiframework.com`), null if can't be obtained from `$_SERVER` and wasn't set.
-     * See [[getHostInfo()]] for security related notes on this property.
```
Comment (cebe) [inline framework/web/Request.php]: why did you remove this? It sensures that the security hint is also contained in the property, not only getter/setter.

**Review by cebe — CHANGES_REQUESTED** [review]
looks good except the docs removal.

File: framework/CHANGELOG.md [hunk 2]
```diff
@@ -57,6 +57,8 @@ Yii Framework 2 Change Log
 - Enh #12988: Changed `textarea` method within the `yii\helpers\BaseHtml` class to allow users to control whether html entities found within `$value` will be double-encoded or not (cyphix333)
 - Enh #13074: Improved `\yii\log\SyslogTarget` with `$options` to be able to change the default `openlog` options. (timbeks)
 - Enh #13050: Added `yii\filters\HostControl` allowing protection against 'host header' attacks (klimov-paul, rob006)
+- Enh #13074: Improved `yii\log\SyslogTarget` with `$options` to be able to change the default `openlog` options. (timbeks)
```
Comment (samdark) [inline framework/CHANGELOG.md]: This is duplicate.

File: framework/CHANGELOG.md [hunk 3]
```diff
@@ -57,6 +57,8 @@ Yii Framework 2 Change Log
...
+- Enh #13050: Added `yii\filters\HostControl` allowing protection against 'host header' attacks (klimov-paul)
```
Comment (samdark) [inline framework/CHANGELOG.md]: This is duplicate as well.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/CHANGELOG.md (hunk 2 above)
Comment (rob006) [inline framework/CHANGELOG.md]: Fixed.

File: framework/CHANGELOG.md (hunk 3 above)
Comment (rob006) [inline framework/CHANGELOG.md]: Fixed.

**Review by cebe — APPROVED** [review]

**Review by samdark — APPROVED** [review]

File: framework/filters/HostControl.php [hunk 4]
```diff
@@ -105,6 +105,13 @@ class HostControl extends ActionFilter
      * host name, creation of absolute URL links, caching page parts and so on.
      */
     public $denyCallback;
+    /**
+     * @var string|null fallback host info (e.g. `http://www.yiiframework.com`) used when [[\yii\web\Request::$hostInfo|Request::$hostInfo]] is invalid.
+     * This value will replace [[\yii\web\Request::$hostInfo|Request::$hostInfo]] before [[$denyCallback]] is called to make sure that
+     * invalid host will not be used for further processing. You can set it to `null` to leave [[\yii\web\Request::$hostInfo|Request::$hostInfo]] untouched.
+     * @see \yii\web\Request::getHostInfo()
```
Comment (SilverFire) [inline framework/filters/HostControl.php]: `@since`

File: framework/filters/HostControl.php [hunk 5]
```diff
@@ -105,6 +105,13 @@ class HostControl extends ActionFilter
      * host name, creation of absolute URL links, caching page parts and so on.
      */
     public $denyCallback;
+    /**
+     * @var string|null fallback host info (e.g. `http://www.yiiframework.com`) used when [[\yii\web\Request::$hostInfo|Request::$hostInfo]] is invalid.
+     * This value will replace [[\yii\web\Request::$hostInfo|Request::$hostInfo]] before [[$denyCallback]] is called to make sure that
+     * invalid host will not be used for further processing. You can set it to `null` to leave [[\yii\web\Request::$hostInfo|Request::$hostInfo]] untouched.
```
Comment (SilverFire) [inline framework/filters/HostControl.php]: Default value should be mentioned

**Review by SilverFire — CHANGES_REQUESTED** [review]
http://www.yiiframework.com/doc-2.0/guide-security-best-practices.html#avoiding-host-header-attacks

Could you update the corresponding guide article?

File: framework/filters/HostControl.php (hunk 5 above)
Comment (dynasource) [inline framework/filters/HostControl.php]: to make sure that invalid host -> that `an` invalid host ...

File: framework/filters/HostControl.php [hunk 6]
```diff
@@ -132,6 +139,11 @@ public function beforeAction($action)
             }
         }
 
+        // replace invalid host info to prevent using it in further processing
```
Comment (dynasource) [inline framework/filters/HostControl.php]: to prevent `from` using

File: tests/framework/filters/HostControlTest.php [hunk 7]
```diff
@@ -131,4 +131,31 @@ public function testDenyCallback()
...
+        $controller = new Controller('test', Yii::$app);
+        $action = new Action('test', $controller);
+        $filter->beforeAction($action);
+
+        $this->assertSame('yiiframework.com', Yii::$app->getRequest()->getHostName());
+    }
+
+    public function testErrorHandlerWithDefaultHost()
+    {
+        $this->setExpectedException('yii\web\NotFoundHttpException', 'Page not found.');
```
Comment (dynasource) [inline tests/framework/filters/HostControlTest.php]: setExpectedException is marked to be deprecated since PHPunit 5.2. Perhabs it better to take this already into account. 
https://github.com/sebastianbergmann/phpunit/wiki/Release-Announcement-for-PHPUnit-5.2.0#changes

**Review by dynasource — CHANGES_REQUESTED** [review]
few minor requests

File: framework/filters/HostControl.php (hunk 4 above)
Comment (cebe) [inline framework/filters/HostControl.php]: the class as a whole is new in 2.0.11, so no need to add that here.

File: framework/filters/HostControl.php (hunk 6 above)
Comment (cebe) [inline framework/filters/HostControl.php]: it's fine as is. 

@yiisoft/core-developers Also I just noticed that for such minor things one can edit the file directly on github and instead of commenting, you can just fix it :)

File: tests/framework/filters/HostControlTest.php (hunk 7 above)
Comment (cebe) [inline tests/framework/filters/HostControlTest.php]: what is the alternative?

Comment (rob006) [inline tests/framework/filters/HostControlTest.php]: @cebe `expectException()` is available in phpunit 4.x?

**Comment by rob006** [comment]
Done.

File: tests/framework/filters/HostControlTest.php (hunk 7 above)
Comment (dynasource) [inline tests/framework/filters/HostControlTest.php]: you can use the expectedException annotation https://phpunit.de/manual/4.8/en/appendixes.annotations.html#appendixes.annotations.expectedException

Comment (rob006) [inline tests/framework/filters/HostControlTest.php]: I'd like to avoid annotations, they are limited and tricky. `setExpectedException()` is already used multiple times in tests, changing this in one place will not make any difference.

**Review by SilverFire — APPROVED** [review]

**Comment by SilverFire** [comment]
Thank you, @rob006
